vignettes/piggyback.Rmd

```diff
@@ -0,0 +1,283 @@
+---
+title: "Piggyback Data atop your GitHub Repository!"
+author: "Carl Boettiger & Tan Ho"
+date: "2023-12-26"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{piggyback}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  results="hide",
+  eval = Sys.getenv("TAN_GH_TOKEN", FALSE)
+)
+
+Sys.setenv(piggyback_cache_duration=0)
+```
+
+## Why `piggyback`?
+
+`piggyback` grew out of the needs of students both in my classroom and in my research 
+group, who frequently need to work with data files somewhat larger than one can 
+conveniently manage by committing directly to GitHub.  As we frequently want to 
+share and run code that depends on >50MB data files on each of our own machines, 
+on continuous integration, and on larger computational servers, data sharing 
+quickly becomes a bottleneck. 
+
+[GitHub allows](https://docs.github.com/en/github/managing-large-files/distributing-large-binaries)
+repositories to attach files of up to 2 GB each to releases as a way to distribute 
+large files associated with the project source code.  There is no limit on the 
+number of files or bandwidth to deliver them.  
+
+## Authentication
+
+No authentication is required to download data from *public* GitHub repositories 
+using `piggyback`. Nevertheless, `piggyback` recommends setting a token when 
+possible to avoid rate limits. To upload data to any repository, or to download 
+data from *private* repositories, you will need to authenticate first. 
+
+`piggyback` uses the same GitHub Personal Access Token (PAT) that devtools, usethis, and
+friends use (`gh::gh_token()`). The current best practice for managing your GitHub
+credentials is detailed in this [usethis vignette](https://usethis.r-lib.org/articles/git-credentials.html).
+
+You can also add the token as an environment variable, which may be useful in 
+situations where you use piggyback non-interactively (i.e.scheduled/automated scripts). 
+Here are the relevant steps:
+
+- Create a [GitHub Token](https://github.com/settings/tokens/new?scopes=repo,gist&description=PIGGYBACK_PAT)
+- Add the environment variable. You can do this:
+  - via project-specific Renviron: `usethis::edit_r_environ("project")`. You should
+  then add the Renviron to your gitignore via `usethis::use_git_ignore(".Renviron")`. 
+  **Avoid committing your GITHUB_PAT to the repository for security reasons!**
+  - via `Sys.setenv(GITHUB_PAT = "{your token}")` in your console for oneoff usage.
+  Avoid adding this line to your R scripts -- remember, the goal here is to avoid 
+  writing your private token in any file that might be shared, even privately.
+
```

Comment:
Streamlined sections "downloading data", "uploading data", "deleting files",  "list files" a bit (I think the remainder after that is mostly the same)